modules/core/04-channel/keeper/upgrade_test.go

```diff
@@ -389,3 +389,98 @@ func (suite *KeeperTestSuite) assertUpgradeError(actualError, expError error) {
 
 	suite.Require().True(errorsmod.IsOf(actualError, expError), actualError)
 }
+
+// TestAbortHandshake tests that when the channel handshake is aborted, the channel state
+// is restored the previous state and that an error receipt is written, and upgrade state which
+// is no longer required is deleted.
+func (suite *KeeperTestSuite) TestAbortHandshake() {
+	var (
+		path *ibctesting.Path
+	)
+
+	tests := []struct {
+		name           string
+		malleate       func()
+		expPass        bool
+		channelPresent bool
+	}{
+		{
+			name:           "success",
+			malleate:       func() {},
+			expPass:        true,
+			channelPresent: true,
+		},
+		{
+			name: "upgrade does not exist",
+			malleate: func() {
+				suite.chainA.DeleteKey(host.ChannelUpgradeKey(path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID))
+			},
+			expPass:        false,
+			channelPresent: true,
+		},
+		{
+			name: "channel does not exist",
+			malleate: func() {
+				suite.chainA.DeleteKey(host.ChannelKey(path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID))
+			},
+			expPass:        false,
+			channelPresent: false,
+		},
+	}
+
+	for _, tc := range tests {
+		tc := tc
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+			path = ibctesting.NewPath(suite.chainA, suite.chainB)
+			suite.coordinator.Setup(path)
+
+			channelKeeper := suite.chainA.GetSimApp().IBCKeeper.ChannelKeeper
+
+			path.EndpointA.ChannelConfig.Version = mock.Version + "-v2"
+
+			suite.Require().NoError(path.EndpointA.ChanUpgradeInit())
+
+			upgradeError := types.NewUpgradeError(1, types.ErrInvalidChannel)
+
+			tc.malleate()
+
+			err := channelKeeper.AbortHandshake(suite.chainA.GetContext(), path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID, upgradeError)
+
+			if tc.expPass {
+				suite.Require().NoError(err)
+
+				channel, found := channelKeeper.GetChannel(suite.chainA.GetContext(), path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID)
+				suite.Require().True(found, "channel should be found")
+
+				suite.Require().Equal(types.OPEN, channel.State, "channel state should be OPEN")
+				suite.Require().Equal(types.NOTINFLUSH, channel.FlushStatus, "channel flush status should be NOTINFLUSH")
+
+				_, found = channelKeeper.GetUpgrade(suite.chainA.GetContext(), path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID)
+				suite.Require().False(found, "upgrade info should be deleted")
+
+				errorReceipt, found := channelKeeper.GetUpgradeErrorReceipt(suite.chainA.GetContext(), path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID)
+				if upgradeError != nil {
+					suite.Require().True(found, "error receipt should be found")
+					suite.Require().Equal(upgradeError.GetErrorReceipt(), errorReceipt, "error receipt should be stored")
```

Comment:
Since presence or not of the error receipt is checked above.

```suggestion
					suite.Require().Equal(upgradeError.GetErrorReceipt(), errorReceipt, "error receipt does not match expected error receipt")
```